### PR TITLE
Bluetooth: Mesh: Deprecate Config Server struct

### PR DESCRIPTION
**Pull request by trond-snekvik**
Encapsulates the Heartbeat state and functionality in a separate
heartbeat module, removing all manipulation of the heartbeat state from
the transport and config server modules.

Adds public documentation for the Heartbeat feature in a separate page
under Bluetooth Mesh.

Moves mesh feature configuration to a separate module, deprecating the
`bt_mesh_cfg_srv` structure. The initial values for the features should
now be enabled through KConfig, where new config entries have been added
for each feature.

This removes the upward dependency on the config server from the core
stack, and makes the config server a pure frontend for the configuration
states, as all spec mandated behavior around the feature states is now
encapsulated.

### CODE REVIEW

**Review by LingaoM — APPROVED** [review]

**Comment by MarcianoPreciado** [comment]
These changes are a huge improvement to the current api but I have a few extra recommendations.
1) Put the new cfg.h setter apis in the mesh.h rather than in subsys/... so that they can easily be included by the user to power their own applications.
2) I'm not certain that the new function `bt_mesh_gatt_proxy_set` will actually fully deactivate or reactive the gatt proxy advertising. In order to make a clean activate/deactivate I think that proxy.c needs some work. It doesn't seem that any of the periodic beaconing functions check whether or not gatt_proxy is enabled. Rather it only gets initiated at the start if it is enabled and doesn't ever check again.

**Comment by MarcianoPreciado** [comment]
**adv.c**
```
static void adv_thread(void *p1, void *p2, void *p3)
{
	BT_DBG("started");

	while (1) {
		struct net_buf *buf;

		if (IS_ENABLED(CONFIG_BT_MESH_PROXY)) {
			buf = net_buf_get(&adv_queue, K_NO_WAIT);
			while (!buf) {
				k_timeout_t timeout;

				timeout = bt_mesh_proxy_adv_start();
				BT_DBG("Proxy Advertising");

				buf = net_buf_get(&adv_queue, timeout);
				bt_mesh_proxy_adv_stop();
			}
		} else {
			buf = net_buf_get(&adv_queue, K_FOREVER);
		}

		if (!buf) {
			continue;
		}

		/* busy == 0 means this was canceled */
		if (BT_MESH_ADV(buf)->busy) {
			BT_MESH_ADV(buf)->busy = 0U;
			adv_send(buf);
		} else {
			net_buf_unref(buf);
		}

		/* Give other threads a chance to run */
		k_yield();
	}
}
```

**Comment by MarcianoPreciado** [comment]
And this is the only place `bt_mesh_gatt_proxy_disable()` is called
**zephyr/subsys/bluetooth/mesh/main.c**
```
void bt_mesh_reset(void)
{
	if (!atomic_test_bit(bt_mesh.flags, BT_MESH_VALID)) {
		return;
	}

	bt_mesh.iv_index = 0U;
	bt_mesh.seq = 0U;

	memset(bt_mesh.flags, 0, sizeof(bt_mesh.flags));

	k_delayed_work_cancel(&bt_mesh.ivu_timer);

	bt_mesh_cfg_reset();

	bt_mesh_rx_reset();
	bt_mesh_tx_reset();

	bt_mesh_net_loopback_clear(BT_MESH_KEY_ANY);

	if (IS_ENABLED(CONFIG_BT_MESH_LOW_POWER)) {
		if (IS_ENABLED(CONFIG_BT_MESH_LPN_SUB_ALL_NODES_ADDR)) {
			uint16_t group = BT_MESH_ADDR_ALL_NODES;

			bt_mesh_lpn_group_del(&group, 1);
		}

		bt_mesh_lpn_disable(true);
	}

	if (IS_ENABLED(CONFIG_BT_MESH_FRIEND)) {
		bt_mesh_friend_clear_net_idx(BT_MESH_KEY_ANY);
	}

	if (IS_ENABLED(CONFIG_BT_MESH_GATT_PROXY)) {
		bt_mesh_proxy_gatt_disable();
	}

	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
		bt_mesh_clear_net();
	}

	(void)memset(bt_mesh.dev_key, 0, sizeof(bt_mesh.dev_key));

	bt_mesh_scan_disable();
	bt_mesh_beacon_disable();

	bt_mesh_comp_unprovision();

	if (IS_ENABLED(CONFIG_BT_MESH_PROV)) {
		bt_mesh_prov_reset();
	}
}
```

**Comment by LingaoM** [comment]
> These changes are a huge improvement to the current api but I have a few extra recommendations.
> 
> 1. Put the new cfg.h setter apis in the mesh.h rather than in subsys/... so that they can easily be included by the user to power their own applications. The only other option would be to make bt_mesh accessible by the user, but it's type definition is hidden away in an include in subsys/.. also.
> 2. I'm not certain that the new function `bt_mesh_gatt_proxy_set` will actually fully deactivate or reactive the gatt proxy advertising. In order to make a clean activate/deactivate I think that proxy.c, adv.c, and main.c need some work. It doesn't seem that any of the periodic beaconing functions check whether or not gatt_proxy is enabled. Rather it only gets initiated at the start if it is enabled and doesn't ever check again.

Hi MarcianoPreciado, thanks you feedback, but i’m want to give you some explanation to why not put cfg.h to application, and why the function proxy disable look like not disable  proxy functionality .

Firstly, cfg.h contains some mesh feature setters api than not allow by call in application layer, because this state must be consistent and be control by stack. Random modifications may cause problems.

Secondly proxy functionality in Mesh spec 1.0.1 has been redefined, as support mesh proxy feature node also advertising proxy beacon even proxy disabled.

Conclusion, I’m agree with trond's consideration, your view is not necessary.

**Comment by MarcianoPreciado** [comment]
@LingaoM Hi Lingao. I see that we come from different perspectives and can possibly learn from one another. I come from the perspective of a mesh-device developer and have product features which are greatly hindered by the current lack of configuration options. I presume you are a maintainer with greater knowledge of the overall zephyr mesh ecosystem and it's requirements, and perhaps you could teach me a thing or two about how it works.

But from my perspective, the zephyr stack makes it impossible to create a product which has the ability to fulfill different roles, such as relay, friend, lpn, etc. Instead, the product can only have one set list of features, and they must always be active. Can you see how this drastically limits the application space for your stack? One would have to sell a product specifically to be a gatt proxy, or specifically a relay, or specifically an lpn. A general device which can occupy multiple roles is an impossibility. A company may not wish to design casings, circuits, and firmwares for 3 products if it can be accomplished with one of each. Physical products are incredibly expensive to design and manufacture. Your view of static devices is a non starter for many would-be users of zephyr. 

Perhaps the inner mechanisms need to be reworked if they are so fragile that they may not be stopped and restarted. Imagine a device which is usually powered but has a battery. If it lost power, the gatt proxy advertisements would drain the battery in less than a few hours. The fact that the application has no way of stopping this from happening is disconcerting. There is nowhere in the mesh specification which suggests a device cannot stop supporting the gatt proxy feature.

**Comment by trond-snekvik** [comment]
Thanks for providing a user perspective on this, @MarcianoPreciado, I really appreciate it.
For APIs, I think user feedback generally weighs heavier than maintainer feedback - this whole exercise is pointless if we're making the wrong product. With that being said, the maintainers ultimately have to decide whether a use case is feasible and fits with the bigger picture.

> Put the new cfg.h setter apis in the mesh.h rather than in subsys/... so that they can easily be included by the user to power their own applications.

I think I agree with this. The usage scenario you're describing definitely should be supported, and I think opening the configuration API will provide the smoothest experience, both internally and externally. @LingaoM raises some valid concerns about the consistency of the configuration state. Configuration clients generally expect to be the "gods" of the networks - if the devices move around too much on their own, it'll be hard to provide a consistent user experience for the network controller devices. In my opinion, it's better to warn about this problem and discourage excessive autonomy in the documentation than block a whole class of user scenarios. The spec doesn't explicitly forbid changing states at runtime, and there are definitely valid reasons to do so. Also, if there are technical reasons why this isn't possible, we should work them out.

> It doesn't seem that any of the periodic beaconing functions check whether or not gatt_proxy is enabled. Rather it only gets initiated at the start if it is enabled and doesn't ever check again.

As @LingaoM says, this is (unfortunately) by design from the spec. Disabling the proxy feature only prevents relaying messages to the rest of the network. Proxy has to be "not supported" for the beaconing to stop, and de-supporting features at runtime is not supported by spec or stack. This is the only feature that acts this way, and hopefully, we'll be able to change it at a later stage, but I think this functionality will have to stay for this PR.

I'll try creating a separate public header for this to see how it fits.

**Comment by LingaoM** [comment]
Looks very nice ! 

I was wondering if it is necessary to add a callback notification to application for the change of role features(such as relay, proxy, friend...).

**Comment by MarcianoPreciado** [comment]
@Trond I see the complication with the GATT proxy in particular. It may be feasible for my product to not include it, and only include PB-GATT features. I'd like to thank you for being open minded. I understand that the work you guys are doing to improve Zephyr isn't always the easiest, and I'm very grateful that you do so anyway.

**Comment by trond-snekvik** [comment]
> I was wondering if it is necessary to add a callback notification to application for the change of role features(such as relay, proxy, friend...).
> For model publish set, model subscription, model bind, model unbind or more, if it necessary to notify application too.
> For friend node, if it necessary to notify application friendship event changes like bt_mesh_lpn_cb.

Perhaps... But I think we should wait until there's a clear use case or someone requests it.

File: include/bluetooth/mesh/cfg_srv.h [hunk 1]
```diff
@@ -93,10 +103,13 @@ struct bt_mesh_cfg_srv {
...
 		 *              @ref BT_MESH_FEAT_FRIEND and
 		 *              @ref BT_MESH_FEAT_LOW_POWER.
+		 *
+		 * @deprecated Please use the @ref bt_mesh_hb_cb registration
+		 *             function.
 		 */
 		void (*func)(uint8_t hops, uint16_t feat);
 	} hb_sub;
-};
+} __deprecated;
```
Comment (trond-snekvik) [inline include/bluetooth/mesh/cfg_srv.h:112]: @jhedberg I'm unsure about the usage of deprecated here. I have marked this struct as deprecated, but it's still accepted in the composition data entry, and the init function still reads the feature flags, potentially overriding the new kconfig entries. Should I instead just flat out remove the structure right away?

Comment (jhedberg) [reply include/bluetooth/mesh/cfg_srv.h:112]: We should probably just remove it (and document it clearly as required by the API change process)

Comment (trond-snekvik) [reply include/bluetooth/mesh/cfg_srv.h]: Done. Also removed the parameter from the `BT_MESH_MODEL_CFG_SRV` macro. This breaks the pattern of having a model struct parameter to these `BT_MESH_MODEL_` macros.

File: include/bluetooth/mesh/cfg.h [hunk 2]
```diff
@@ -35,6 +35,159 @@ enum bt_mesh_feat_state {
...
+ *  @see BT_MESH_TRANSMIT
+ *
+ *  @param relay New Relay feature state.
+ *  @param xmit  New Relay retransmit parameters. Use @ref BT_MESH_TRANSMIT for
+ *               encoding.
+ *
+ *  @retval 0        Successfully changed the Relay configuration.
+ *  @retval -ENOTSUP The Relay feature is not supported.
+ */
+int bt_mesh_relay_set(bool relay, uint8_t xmit);
```
Comment (joerchan) [inline include/bluetooth/mesh/cfg.h:128]: Why is relay an bool here, and an enum bt_mesh_feat_state  in the get function?

Same question goes for the next following proxy and friendship etc. So it must be intentional, but it looks really strange to me.

File: include/bluetooth/mesh/heartbeat.h [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	 * Bitmap of features that trigger a Heartbeat publication if
+	 * they change. Legal values are @ref BT_MESH_FEAT_RELAY,
+	 * @ref BT_MESH_FEAT_PROXY, @ref BT_MESH_FEAT_FRIEND and
+	 * @ref BT_MESH_FEAT_LOW_POWER.
+	 */
+	uint16_t feat;
+	/** Network index used for publishing. */
+	uint16_t net_idx;
+	/** Publication interval in seconds. */
+	uint32_t period;
```
Comment (joerchan) [inline include/bluetooth/mesh/heartbeat.h:44]: Should this one be called interval?

File: include/bluetooth/mesh/heartbeat.h [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+};
+
+/** Heartbeat Subscription parameters. */
+struct bt_mesh_hb_sub {
+	/** Total subscription time in seconds. */
+	uint32_t period;
```
Comment (joerchan) [inline include/bluetooth/mesh/heartbeat.h:50]: total perhaps? or update documentation: "Subscription period in seconds"? Or "Subscription period (total time) in seconds"

File: subsys/bluetooth/mesh/Kconfig [hunk 5]
```diff
@@ -107,13 +107,21 @@ config BT_MESH_PB_GATT
 	  GATT.
 
 config BT_MESH_GATT_PROXY
-	bool "GATT Proxy Service"
+	bool "GATT Proxy Service support"
```
Comment (joerchan) [inline subsys/bluetooth/mesh/Kconfig:110]: GATT Proxy Service feature?

File: subsys/bluetooth/mesh/cfg.c [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+	    atomic_test_bit(bt_mesh.flags, BT_MESH_VALID)) {
+		bt_mesh_store_cfg();
+	}
+
+	return 0;
+}
+
+enum bt_mesh_feat_state bt_mesh_relay_get(void)
+{
+	return atomic_test_bit(bt_mesh.flags, BT_MESH_RELAY);
```
Comment (joerchan) [inline subsys/bluetooth/mesh/cfg.c:171]: Return value of atomic_test_bit is bool, it looks wrong to have enum bt_mesh_feat_state as return value.

File: include/bluetooth/mesh/cfg.h (hunk 2 above)
Comment (trond-snekvik) [reply include/bluetooth/mesh/cfg.h:128]: It is intentional, but I think the intention might not be very good 😄 

These values can only be set to enabled/disabled at runtime, but can be set to "not supported" by the compile time configuration. Coming back to it now, I definitely agree it looks jarring. I'll change it to use the feature enum, and return an error if the value is invalid.

File: include/bluetooth/mesh/heartbeat.h (hunk 3 above)
Comment (trond-snekvik) [reply include/bluetooth/mesh/heartbeat.h:44]: The heartbeat parameter names are unchanged from master, and they match the spec. I'd rather change the documentation to match the parameter terminology, so we don't move away from the spec naming

File: subsys/bluetooth/mesh/Kconfig (hunk 5 above)
Comment (trond-snekvik) [reply subsys/bluetooth/mesh/Kconfig:110]: "support" is used for the rest of the mesh features, I'd prefer to stay consistent

File: subsys/bluetooth/mesh/cfg.c (hunk 6 above)
Comment (trond-snekvik) [reply subsys/bluetooth/mesh/cfg.c:171]: This is a similar problem to the set API bool, I'll convert it.

**Review by jhedberg — APPROVED** [review]

**Review by joerchan — APPROVED** [review]

**Pull request merged**